hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/metadata/SigningCertFromMetadataExtractor.java

```diff
@@ -0,0 +1,59 @@
+package uk.gov.ida.hub.samlengine.metadata;
+
+import net.shibboleth.utilities.java.support.component.ComponentInitializationException;
+import net.shibboleth.utilities.java.support.resolver.CriteriaSet;
+import net.shibboleth.utilities.java.support.resolver.ResolverException;
+import org.opensaml.core.criterion.EntityIdCriterion;
+import org.opensaml.saml.common.xml.SAMLConstants;
+import org.opensaml.saml.criterion.EntityRoleCriterion;
+import org.opensaml.saml.criterion.ProtocolCriterion;
+import org.opensaml.saml.metadata.resolver.MetadataResolver;
+import org.opensaml.saml.saml2.metadata.SPSSODescriptor;
+import org.opensaml.saml.security.impl.MetadataCredentialResolver;
+import org.opensaml.security.credential.Credential;
+import org.opensaml.security.credential.UsageType;
+import org.opensaml.security.credential.criteria.impl.EvaluablePublicKeyCredentialCriterion;
+import org.opensaml.security.criteria.UsageCriterion;
+import org.opensaml.security.x509.X509Credential;
+import uk.gov.ida.hub.samlengine.exceptions.CertificateForCurrentPrivateSigningKeyNotFoundInMetadataException;
+import uk.gov.ida.hub.samlengine.exceptions.UnableToResolveSigningCertsForHubException;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.security.PublicKey;
+import java.security.cert.X509Certificate;
+
+import static uk.gov.ida.hub.samlengine.SamlEngineModule.VERIFY_METADATA_RESOLVER;
+
+public class SigningCertFromMetadataExtractor {
+
+    private final MetadataCredentialResolver credentialResolver;
+    private final String hubEntityId;
+
+    @Inject
+    public SigningCertFromMetadataExtractor(@Named(VERIFY_METADATA_RESOLVER) MetadataResolver metadataResolver,
+                                            @Named("HubEntityId") String hubEntityId) throws ComponentInitializationException {
+        this.credentialResolver = new MetadataCredentialResolverInitializer(metadataResolver).initialize();
+        this.hubEntityId = hubEntityId;
+    }
+
+    public X509Certificate getSigningCertForCurrentSigningKey(PublicKey publicSigningKey) {
+        CriteriaSet criteriaSet = new CriteriaSet(
+            new EntityIdCriterion(hubEntityId),
+            new UsageCriterion(UsageType.SIGNING),
+            new ProtocolCriterion(SAMLConstants.SAML20P_NS),
+            new EntityRoleCriterion(SPSSODescriptor.DEFAULT_ELEMENT_NAME),
+            new EvaluablePublicKeyCredentialCriterion(publicSigningKey)
+        );
+        try {
+            Credential credential = credentialResolver.resolveSingle(criteriaSet);
+            if (credential instanceof X509Credential) {
+                return ((X509Credential)credential).getEntityCertificate();
```

Comment:
Good call, I'll do that.

Reply:
That's a good addition.  It would be nice to test the CriteriaSet in some way, but perhaps that isn't possible.  Are there eidas tests that would fail if the correct signing cert was not found in the metadata?

Reply:
Stub country has a runtime check that the signing cert is in the Authn request KeyInfo and we have code in saml engine which checks that the public key in the certificate is the same as the public key derived from the private key here: https://github.com/alphagov/verify-hub/pull/154/files#diff-88e1cfadc6ee3e788caa52b0b8836223R52

Reply:
However, production is the only environment which has multiple signing certificates in the metadata

Reply:
That's why I think it should be tested - if it goes wrong the place it is most likely to be discovered is prod (sadface)

There was [previously] no test in this repo that checked the outbound message had the correct cert in KeyInfo, and that would be a good proxy to check that this code is working...  

This PR exists because the cert was removed then had to be re-added after the issue was discovered in a prod-like environment - it would be good to know that this has broken as early as possible

Reply:
The CriteriaSet is the thing that sifts the right bit out of the metadata, and that is not tested right now, but ultimately that might not be very easy so :shrug: